lib/leadership/watcher.go

```diff
@@ -0,0 +1,159 @@
+package leadership
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+	etcd "go.etcd.io/etcd/clientv3"
+	v3rpc "go.etcd.io/etcd/etcdserver/api/v3rpc/rpctypes"
+)
+
+// NewWatcher returns a new watcher to relay changes to the specified key.
+// The watcher is automatically active.
+func NewWatcher(config WatcherConfig) (*Watcher, error) {
+	if err := config.checkAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	w := &Watcher{
+		config:   config,
+		ctx:      ctx,
+		cancel:   cancel,
+		respChan: make(chan etcd.Event),
+	}
+	go w.loop()
+	return w, nil
+}
+
+// Watcher can follow and push notifications whenever
+// there is a change for the specified key.
+type Watcher struct {
+	config   WatcherConfig
+	respChan chan etcd.Event
+	ctx      context.Context
+	cancel   context.CancelFunc
+}
+
+// WatcherConfig defines watcher configuration
+type WatcherConfig struct {
+	// Key specifies the key to watch
+	Key string
+	// Client specifies the etcd client
+	Client *etcd.Client
+	// RetryTimeout specifies the timeout between reconnect attempts
+	RetryTimeout time.Duration
+	// FieldLogger specifies the logger
+	logrus.FieldLogger
+	clock clockwork.Clock
+}
+
+// Stop requests watcher to stop and release resources
+func (r *Watcher) Stop() {
+	r.cancel()
+	for range r.respChan {
+	}
+}
+
+// RespChan returns the channel that the watcher is relaying the updates to.
+// The channel must be served to avoid blocking the watcher
+func (r *Watcher) RespChan() <-chan etcd.Event {
+	return r.respChan
+}
+
+func (r *Watcher) loop() {
+	defer close(r.respChan)
+
+	ticker := r.config.clock.NewTicker(r.config.RetryTimeout)
+	defer ticker.Stop()
+	opts := []etcd.OpOption{
+		etcd.WithPrefix(),
+	}
+	var revision int64
+
+	resp, err := r.config.Client.Get(r.ctx, r.config.Key, opts...)
+	if err == nil && len(resp.Kvs) != 0 {
+		// Update the receiver
+		select {
+		case r.respChan <- etcd.Event{
+			Kv: resp.Kvs[0],
+		}:
+			revision = resp.Kvs[0].ModRevision + 1
+		case <-r.ctx.Done():
+			return
+		}
+	}
+
+	wc := etcd.NewWatcher(r.config.Client)
```

Comment:
Just incase this is meant to be included, my understanding is a Watch should actually be linked to a concurrency.Session so that the Session can detect failure scnearios the Watch itself can't detect.

Reply:
Reworked to use session.